File: Templates/Day_4/DerivedDatatypes/SampleSolutions/C/reductionExample_sol.c
#include <stdio.h>
#include <math.h>
#include "mpi.h"
#include "realint_pair.h"


void setup_mpi(MPI_Comm* pcomm, int* psize, int* prank)
{
    MPI_Init(NULL, NULL);
    
    *pcomm = MPI_COMM_WORLD;
    
    MPI_Comm_size(*pcomm, psize);
    MPI_Comm_rank(*pcomm, prank);
    
    return;
}

void addmax_ri(void* pfirst, void* psecond, int* plen,
	       MPI_Datatype* ptype)
{
  ri_pair* pf = (ri_pair*)pfirst;
  ri_pair* ps = (ri_pair*)psecond;

  for (int i=0; i < *plen; i++)
    {
      ps[i].rpart += pf[i].rpart;
      ps[i].ipart = fmax(pf[i].ipart, ps[i].ipart);
    }
  return;
}


int main(int argc, char **argv) 
{

  // setting up MPI

  MPI_Comm my_world;
  int my_rank, world_size;

  setup_mpi(&my_world, &world_size, &my_rank);

  ri_pair my_values, received_values, reduced_values;

  my_values.rpart = my_rank;
  my_values.ipart = my_rank;

  printf("On rank %d my values are: (%f, %d)\n", 
	 my_rank, my_values);
 
  // Stuff to make derived datatype
  const int nb_entries = 2;
  int bleng[nb_entries];
  bleng[0] = 1;
  bleng[1] = 1;

  MPI_Aint start_address, disp[nb_entries];
  MPI_Get_address(&my_values,       &start_address);
  MPI_Get_address(&my_values.rpart, &(disp[0]));
  MPI_Get_address(&my_values.ipart, &(disp[1]));

  for (int i=0; i < nb_entries; i++)
    disp[i] = disp[i] - start_address;

  MPI_Datatype types[nb_entries]; 
  types[0] = MPI_DOUBLE;
  types[1] = MPI_INT;

  MPI_Datatype ri_mpitype;
  MPI_Type_create_struct(nb_entries, bleng, disp, types, &ri_mpitype);
  MPI_Type_commit(&ri_mpitype);

  if (my_rank == 0)
    {
      MPI_Status stat;
      MPI_Recv(&received_values, 1, ri_mpitype, 1, 1, my_world, &stat);
      printf("On rank 0 we received from 1: (%f, %d)\n", 
	     received_values);
    }
  else if(my_rank == 1)
    {
      MPI_Send(&my_values, 1, ri_mpitype, 0, 1, my_world);
    }

  // reduction operator
  MPI_Op ri_addmax_mpiop;
  MPI_Op_create(&addmax_ri, 1, &ri_addmax_mpiop);

  MPI_Reduce(&my_values, &reduced_values, 1, ri_mpitype, 
	     ri_addmax_mpiop, 0, my_world);

  if (my_rank == 0)
    {
      printf("On rank 0 we got reduced values: (%f, %d)\n", 
	     reduced_values);
    }
	     

 
  MPI_Finalize();

  return 0;
}

File: Templates/Day_4/DerivedDatatypes/SampleSolutions/Fortran/reductionExample_sol.f90
Program ReductionTest

  use mpi
  use realint_pair

  Implicit None

  type(ri_pair) :: my_values, reduced_values

  Integer :: my_rank, world_size
  Integer(kind=mpi_address_kind) :: start_address, secnd_address

  Integer, parameter :: nb_entries=2
  Integer, dimension(nb_entries) :: bleng, types
  Integer(kind=mpi_address_kind),dimension(nb_entries) :: displ

  Integer :: ri_mpitype, ri_addmax_mpiop

  Integer, dimension(Mpi_Status_size) :: mstatus
  Integer :: merror

  Call MPI_Init(merror)

  Call MPI_Comm_rank(MPI_Comm_world, my_rank, merror)
  Call MPI_Comm_size(MPI_Comm_world, world_size, merror)

  my_values%realpart = my_rank
  my_values%intpart  = my_rank

  print *,"On rank", my_rank," my_values contains", my_values

  ! setting up an MPI derived data type
  Call MPI_Get_address(my_values%realpart, start_address, merror)
  Call MPI_Get_address(my_values%intpart,  secnd_address, merror)

  bleng = 1

  displ(1) = 0
  displ(2) = secnd_address - start_address

  types(1) = MPI_double_precision
  types(2) = MPI_integer

  print *,"On rank", my_rank," displacements are:", displ

  Call MPI_Type_create_struct(nb_entries, bleng, displ, types, ri_mpitype, &
       merror)
  Call MPI_Type_commit( ri_mpitype, merror)

  If (my_rank == 1) then
     Call MPI_Send(my_values, 1, ri_mpitype, 0, 0, MPI_Comm_world, merror)
  elseif (my_rank == 0) then
     Call MPI_Recv(reduced_values, 1, ri_mpitype, 1, 0, MPI_comm_world, &
          mstatus, merror)
     print *,"On rank", my_rank," received value: ", reduced_values
  endif


  Call MPI_op_create(addmax_ri, .true., ri_addmax_mpiop, merror)

  Call MPI_Reduce(my_values, reduced_values, 1, ri_mpitype, &
       ri_addmax_mpiop, 0, MPI_Comm_world, merror)
  if (my_rank == 0) then
     print *,"On rank", my_rank," reduced_values after reduce: ", reduced_values
  endif

  Call MPI_Finalize(merror)

End Program ReductionTest


